packages/serverpod/lib/src/server/server.dart

```diff
@@ -132,9 +147,18 @@ class Server implements RouterInjectable {
     if (serverpod.config.loggingMode == ServerpodLoggingMode.verbose) {
       router.use('/', _verboseLogging);
     }
+
+    // Register core middleware first to ensure they wrap all user middleware
     router
       ..use('/', _headers)
-      ..use('/', _reportException)
+      ..use('/', _reportException);
+
+    // Register user-provided middleware
+    for (var middleware in _middleware) {
+      router.use('/', middleware);
+    }
+
+    router
```

Comment:
This is fine, but just for clarification. You could also loop over the middleware after adding the handlers. Middleware is applied by hierarchy first, then order, meaning that even though you call `use` after `add` (`get`,  etc.) it will still be applied to all handlers below that path.

The order is only important when attaching middleware at the same path. In that case they are composed first-to-last as outer-to-inner.

In particular the middleware is also applied to `_health`, `/websocket`, `/v1/websoket`, `/serverpod_cloud_storage`, and `/**` (ie. the serverpod endpoints).

Reply:
Do you see anything possibly wrong if the middlewares runs for those special endpoints, especially for the websocket ones?

Reply:
No, I think it is okay. You can use middleware on web-socket endpoints to drop connections before they are started fx. We just need to make sure the docs are not incorrect.